doc/bootloader/README.md

```diff
@@ -0,0 +1,113 @@
+
+### The bootloader
+[The bootloader](https://github.com/InfiniTimeOrg/pinetime-mcuboot-bootloader) is based on [MCUBoot, a secure bootloader for 32 bits microcontrollers](https://github.com/mcu-tools/mcuboot). It was originally written by [Lup Yuen](https://github.com/lupyuen).
+
+MCUBoot is run at boot time. In normal operation, it just jumps to the reset handler of the application firmware to run it. Once the application firmware is running, MCUBoot does not run at all.
+
+![MCUBoot boot sequence diagram](boot.png)
+
+But MCUBoot does much more than that : it can upgrade the firmware that is currently running by a new one, and it is also able to revert to the previous version of the firmware in case the new one does not run properly.
+
+To do this, it uses 2 memory 'slots' :
+- **The primary slot** (located in the internal flash memory of the nRF52832) : it contains the current firmware, the one that will be executed by MCUBoot
+- **The secondary slot** (located in the external SPI flash memory) : it is used to store the upgraded version of the firmware, when available.
+
+At boot time, MCUBoot detects that a new version of the firmware is available in the secondary slot and swaps them : the current version of the firmware is copied from the primary to the secondary slot and vice-versa.
+
+When the swapping is done, it runs the new version of the firmware from the primary slot.
+
+![MCUBoot update sequence diagram](upgrade.png)
+
+The next time MCUBoot will be run (after an MCU reset), MCUBoot will check if the new firmware ran correctly (it must set a flag in memory). If it is not the case, it'll revert to the previous version of the firmware by copying it from the secondary to the primary.
+
+![MCUBoot update sequence diagram](recover.png)
+
+You'll find more information about MCUBoot, its design and how it works in [the documentation of the project](https://docs.mcuboot.com/).
+
+The [pinetime-mcuboot-bootloader](https://github.com/InfiniTimeOrg/pinetime-mcuboot-bootloader) integrates MCUBoot, provides low-level drivers for the display and SPI flash memory and implements a very basic UI. See [the documentation of the project for more details about this](https://github.com/InfiniTimeOrg/pinetime-mcuboot-bootloader#boot-flow).
+
+An application firmware must implement the following functionalities to be compatible with our bootloader:
+
+- The application firmware will be flashed at **address 0x8000 of the internal flash**. It must then be **linked** with a start address of 0x8000 (instead of 0x00 by default).
+- The bootloader starts the **watchdog** just before loading the application firmware. The application firmware must **refresh the watchdog regularly**. If the watchdog is not refreshed in time, the microcontroller will reset, and the bootloader will consider the firmware as unfunctional and will revert to the previous one.
```

Comment:
suggesting `nonfunctional` from my computer dictionary, and `micro-controller`

```suggestion
- The bootloader starts the **watchdog** just before loading the application firmware.
  The application firmware must **refresh the watchdog regularly**.
  If the watchdog is not refreshed in time, the micro-controller will reset, and the bootloader will consider the firmware as nonfunctional and will revert to the previous one.
```

Reply:
I think it should be `microcontroller`. See [this Wikipedia article](https://en.wikipedia.org/wiki/Microcontroller).

Reply:
Switch to requirement first, rationale second. Explicitly state what will be reverted.

Question: Does the bootloader take into account the reboot reason, or is it just because the validation bit isn't set that it would revert the image? Would the firmware reversion still happen if the WDT fired some time after the validation bit was set?

Assuming it's just about the WDT firing before the validation is set:

```suggestion
- The application firmware must **refresh the watchdog regularly**.
  The bootloader starts the watchdog just before loading the application firmware.
  If the watchdog is not refreshed in time, the microcontroller will reset.
  After the reset the bootloader will notice that the validation bit has not been set, and will revert to the previous application firmware image.
```